pandas/core/algorithms.py

```diff
@@ -728,25 +733,31 @@ def factorize(
 
     if not isinstance(values.dtype, np.dtype):
         # i.e. ExtensionDtype
+        assert dropna or sort
```

Comment:
can you add a comment about why this assertion should hold

Reply:
Ah, this was just to demonstrate no tests hit here; will be adding tests that hit here with both dropna and sort being False. See OP for details.